File: api/mcfrs_api/services/change_detection.py
"""Compare two snapshots of the active set. Pure. An incident absent from the
new snapshot is CLOSED (the upstream active list dropped it), never deleted."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime

from mcfrs_api.models import Incident, IncidentStatus


@dataclass
class ChangeSet:
    added: list[Incident] = field(default_factory=list)
    changed: list[Incident] = field(default_factory=list)
    closed: list[Incident] = field(default_factory=list)
    unchanged: int = 0

    @property
    def total(self) -> int:
        return len(self.added) + len(self.changed) + len(self.closed)


def fingerprint(i: Incident) -> str:
    """Fields whose change is meaningful to clients (raw_payload/updated_at churn is ignored)."""
    units = ",".join(sorted(f"{u.unit}:{u.status or ''}" for u in i.units))
    parts = [
        i.status,
        i.call_type,
        i.call_subtype or "",
        i.priority or "",
        i.address or "",
        i.latitude,
        i.longitude,
        i.station_area,
        i.alarm_level or "",
        i.talkgroup or "",
        i.channel or "",
        units,
    ]
    return "|".join("" if p is None else str(p) for p in parts)


def detect_changes(previous: dict[str, Incident], nxt: list[Incident], now: datetime) -> ChangeSet:
    cs = ChangeSet()
    seen: set[str] = set()
    for n in nxt:
        seen.add(n.id)
        p = previous.get(n.id)
        if p is None:
            cs.added.append(n)
        elif fingerprint(p) != fingerprint(n):
            cs.changed.append(n.model_copy(update={"updated_at": n.updated_at or now}))
        else:
            cs.unchanged += 1
    for pid, p in previous.items():
        if pid not in seen and p.status != IncidentStatus.CLOSED:
            cs.closed.append(
                p.model_copy(
                    update={"status": IncidentStatus.CLOSED, "closed_at": p.closed_at or now, "updated_at": now}
                )
            )
    return cs

File: src/services/changeDetection.ts
import type { Incident } from "@/types/incident";

/**
 * Compare two snapshots of the active set. Pure. An incident absent from the new
 * snapshot is CLOSED (the upstream active list dropped it), never deleted.
 */
export interface ChangeSet {
  added: Incident[];
  changed: Incident[];
  closed: Incident[];
  unchanged: number;
}

/** Fields whose change is meaningful to clients (rawPayload/updatedAt churn is ignored). */
function fingerprint(i: Incident): string {
  const units = i.units.map((u) => `${u.unit}:${u.status ?? ""}`).sort().join(",");
  return [i.status, i.callType, i.callSubtype ?? "", i.priority ?? "", i.address ?? "", i.latitude ?? "", i.longitude ?? "",
    i.stationArea ?? "", i.alarmLevel ?? "", i.talkgroup ?? "", i.channel ?? "", units].join("|");
}

export function detectChanges(previous: ReadonlyMap<string, Incident>, next: readonly Incident[], now: Date): ChangeSet {
  const added: Incident[] = [];
  const changed: Incident[] = [];
  const seen = new Set<string>();
  let unchanged = 0;
  for (const n of next) {
    seen.add(n.id);
    const p = previous.get(n.id);
    if (!p) added.push(n);
    else if (fingerprint(p) !== fingerprint(n)) changed.push({ ...n, updatedAt: n.updatedAt ?? now });
    else unchanged++;
  }
  const closed: Incident[] = [];
  for (const [id, p] of previous) {
    if (!seen.has(id) && p.status !== "closed") closed.push({ ...p, status: "closed", closedAt: p.closedAt ?? now, updatedAt: now });
  }
  return { added, changed, closed, unchanged };
}
